File: backend-spring/src/main/java/io/github/agussyahrilmubarok/backend/delivery/http/rest/GlobalExceptionHandler.java
package io.github.agussyahrilmubarok.backend.delivery.http.rest;

import io.github.agussyahrilmubarok.backend.common.exception.*;
import io.github.agussyahrilmubarok.backend.delivery.http.model.ApiResponse;
import java.util.Map;
import java.util.stream.Collectors;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(EmailAlreadyInUseException.class)
    public ResponseEntity<ApiResponse<Object>> handleEmailAlreadyExistsException(EmailAlreadyInUseException ex) {
        return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(new ApiResponse<>("Conflict", Map.of("email", ex.getMessage())));
    }

    @ExceptionHandler(EmailNotRegisteredException.class)
    public ResponseEntity<ApiResponse<Object>> handleEmailNotRegisteredException(EmailNotRegisteredException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(new ApiResponse<>("Unauthorized", Map.of("email", ex.getMessage())));
    }

    @ExceptionHandler(JwtExpiredTokenException.class)
    public ResponseEntity<ApiResponse<Object>> handleJwtExpiredTokenException(JwtExpiredTokenException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(new ApiResponse<>("Unauthorized", Map.of("error", ex.getMessage())));
    }

    @ExceptionHandler(JwtInvalidTokenException.class)
    public ResponseEntity<ApiResponse<Object>> handleJwtInvalidTokenException(JwtInvalidTokenException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(new ApiResponse<>("Unauthorized", Map.of("error", ex.getMessage())));
    }

    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<ApiResponse<Object>> handleNotFoundException(NotFoundException ex) {
        return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(new ApiResponse<>("Not found", Map.of("error", ex.getMessage())));
    }

    @ExceptionHandler(PasswordMismatchException.class)
    public ResponseEntity<ApiResponse<Object>> handleWrongPasswordException(PasswordMismatchException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(new ApiResponse<>("Unauthorized", Map.of("password", ex.getMessage())));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiResponse<Object>> handleValidationException(MethodArgumentNotValidException ex) {
        Map<String, String> errors = ex.getBindingResult().getFieldErrors().stream()
                .collect(Collectors.toMap(
                        FieldError::getField,
                        field -> field.getDefaultMessage() != null ? field.getDefaultMessage() : "Invalid value",
                        (existing, duplicate) -> existing));
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ApiResponse<>("Validation failed", errors));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiResponse<Object>> handleException(Exception ex) {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>("Internal server error", Map.of("error", "Something went wrong")));
    }
}

File: backend-dotnet/Backend/Delivery/Http/Middleware/GlobalExceptionHandler.cs
using System.Net;
using System.Text.Json;
using Backend.Common.Exceptions;
using Backend.Delivery.Http.Payload;

namespace Backend.Delivery.Http.Middleware;

public class GlobalExceptionHandler : IMiddleware
{
    private readonly ILogger<GlobalExceptionHandler> _logger;

    public GlobalExceptionHandler(ILogger<GlobalExceptionHandler> logger)
    {
        _logger = logger;
    }

    public async Task InvokeAsync(HttpContext context, RequestDelegate next)
    {
        try
        {
            await next(context);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Unhandled exception: {Message}", ex.Message);
            await HandleExceptionAsync(context, ex);
        }
    }

    private static async Task HandleExceptionAsync(HttpContext context, Exception exception)
    {
        context.Response.ContentType = "application/json";

        ApiResponse<object> response;

        switch (exception)
        {
            case NotFoundException:
                context.Response.StatusCode = (int)HttpStatusCode.NotFound;
                response = new ApiResponse<object>(
                    "Not found.",
                    new Dictionary<string, string>
                    {
                        ["error"] = exception.Message
                    });
                break;

            case EmailAlreadyExistsException:
                context.Response.StatusCode = (int)HttpStatusCode.Conflict;
                response = new ApiResponse<object>(
                    "Conflict.",
                    new Dictionary<string, string>
                    {
                        ["email"] = exception.Message
                    });
                break;

            case JwtExpiredTokenException:
                context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                response = new ApiResponse<object>(
                    "Unauthorized.",
                    new Dictionary<string, string>
                    {
                        ["error"] = exception.Message
                    });
                break;

            case JwtInvalidTokenException:
                context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                response = new ApiResponse<object>(
                    "Unauthorized.",
                    new Dictionary<string, string>
                    {
                        ["error"] = exception.Message
                    });
                break;

            case EmailNotRegisteredException:
                context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                response = new ApiResponse<object>(
                    "Unauthorized.",
                    new Dictionary<string, string>
                    {
                        ["email"] = exception.Message
                    });
                break;

            case WrongPasswordException:
                context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                response = new ApiResponse<object>(
                    "Unauthorized.",
                    new Dictionary<string, string>
                    {
                        ["password"] = exception.Message
                    });
                break;

            case UnauthorizedAccessException:
                context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                response = new ApiResponse<object>("Unauthorized.",
                    new Dictionary<string, string> { ["error"] = exception.Message });
                break;

            case AppValidationException ve:
                context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
                response = new ApiResponse<object>(
                    "Validation failed.",
                    ve.Errors);
                break;

            default:
                context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
                response = new ApiResponse<object>(
                    "Something went wrong.",
                    new Dictionary<string, string>
                    {
                        ["error"] = "Something went wrong."
                    });
                break;
        }

        var json = JsonSerializer.Serialize(response, new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        });

        await context.Response.WriteAsync(json);
    }
}